composer/core/state.py

```diff
@@ -670,25 +715,28 @@ def state_dict(self) -> Dict[str, Any]:
             elif attribute_name == 'model':
                 # Save model directly instead of by class name, since model may be wrapped by DistributedDataParallel
                 # If it is DDP wrapped, do not save the `module.` prefix, as that is an implementation detail
-                with get_fsdp_rank0_cpu_save_context(
-                        attribute_value) if self.fsdp_enabled else contextlib.nullcontext():
+
+                with fsdp_state_dict_type_context(
+                        attribute_value, state_dict_type=self.fsdp_config['state_dict_type']) if (
+                            self.fsdp_enabled and self.fsdp_config is not None) else contextlib.nullcontext():
                     model_state = attribute_value.state_dict()
 
                 if self.is_model_ddp:
                     torch.nn.modules.utils.consume_prefix_in_state_dict_if_present(model_state, 'module.')
                 serialized_value = model_state
             elif attribute_name == 'optimizers':
-                if self.fsdp_enabled:
-                    serialized_value = {}
-                    for obj in ensure_tuple(attribute_value):
-                        serialized_value = {
-                            type(obj).__qualname__:
-                                get_fsdp_full_optim_state_dict(model=self.model, optim=obj, rank0_only=True)
-                        }
-                else:
-                    serialized_value = {
-                        type(obj).__qualname__: obj.state_dict() for obj in ensure_tuple(attribute_value)
+                optimizer = ensure_tuple(attribute_value)[
```

Comment:
Could you raise an error here if a user does pass in more than one optimizer? At some point we should nuke the > 1 optimizer code...